### PR TITLE
sql/pgwire: populate table and column ID in RowDescription

### LINKED ISSUE #44161: pgwire, sql: pgjdbc ResultSet.insertRow fails due to missing Table OID
**Describe the problem**

We are trying to migrate an application based on the Axon Framework. Within this framework we've configured the `JdbcTokenStore` to handle reading and creating tokens for event processing. When we run this on CockroachDB, an obscure error pops up: `ERROR: Column "" not found`.

It took us some time to figure out where the `Column ""` was coming from, but we found it. The `JdbcTokenStore` adds a token to the database by using the following snippet:

```
resultSet.moveToInsertRow();
resultSet.updateObject(schema.tokenColumn(), token == null ? null : entry.getSerializedToken().getData());
resultSet.updateString(schema.tokenTypeColumn(),
        token == null ? null : entry.getSerializedToken().getType().getName());
resultSet.updateString(schema.timestampColumn(), entry.timestampAsString());
resultSet.updateString(schema.ownerColum(), entry.getOwner());
resultSet.updateString(schema.processorNameColumn(), processorName);
resultSet.updateInt(schema.segmentColumn(), segment);
resultSet.insertRow();
```

The error occurs on the last line of the snippet. We've traced the error to the `PgResultSet` from the postgresql JDBC-driver, which actually creates the statement `INSERT INTO TokenEntry("") VALUES ($1)`. This baffled us for a bit so we went on debugging and found the following piece of code hidden inside `PgResultSetMetaData`:

```
  public String getBaseColumnName(int column) throws SQLException {
    Field field = getField(column);
    if (field.getTableOid() == 0) {
      return "";
    }
    fetchFieldMetaData();
    FieldMetadata metadata = field.getMetadata();
    return metadata == null ? "" : metadata.columnName;
  }
```

While debugging, the `field.getTableOid()` always returns `0`, causing all field names to be resolved to the empty string (`""`). This in turn cause the `HashMap<String, Object> updateValues` in `PgResultSet` to have exactly 1 entry `<"", 0>`, which is the result of the last `resultSet.updateXXX` in the snippet up on top of this issue. This then creates the invalid statement `INSERT INTO TokenEntry("") VALUES ($1)` and makes the whole thing break.

**To Reproduce**

See above.

**Expected behavior**

Apparently, the `pgjdbc`-driver expects/requires all field-metadata to include a non-zero value for Table OID when it is reading the resultset-metadata. (This happens inside `QueryExecutorImpl.receiveFields` on line 2456). 

From what I could find in the Cockroach internal schemas and `pg_*` tables, there are no OIDs assigned to tables, which is probably why this value is `0` for all tables. 

I would expect the `ResultSet.insertRow` to behave properly.

**Environment:**
 - CockroachDB version v19.2.0
 - Server OS: Official Docker container
 - Client app: JDBC using version 42.1.4 of `org.postgresql:postgresql`

**Additional context**

I have created a local fork of the Postgresql JDBC driver and removed the if-statement to see if things would work normally after this. I can confirm that database operations perform as expected when this is done. So as far as I can see, there are no other issues holding back `ResultSet.insertRow`.

### PR DESCRIPTION
**Pull request by rafiss**
The RowDescription in the wire protocol contains metadata that points to
the source of the returned column. Drivers such as pgjdbc rely on this
metadata in order for certain functionality to work.

Release note (sql change): The RowDescription message of the pgwire
protocol now contains the table OID and column ID for each column in the
result set. These values correspond to pg_attribute.attrelid and
pg_attribute.attnum. If a result column does not refer to a simple table
or view, these values will be zero, as they were before.

fixes #44161

### CODE REVIEW

File: pkg/sql/pgwire/testdata/pgtest/int_size [hunk 1]
```diff
@@ -92,6 +93,6 @@ Query {"String": "SELECT * FROM t1"}
 until
 ReadyForQuery
 ----
```
Comment (jordanlewis) [inline pkg/sql/pgwire/testdata/pgtest/int_size:95]: Could you add a couple more tests in here that check what happens when a column gets threaded through a few operators?

I'm thinking we should make sure that, for example, something like `SELECT c FROM t JOIN t2 ON ... WHERE ... ORDER BY d` still tracks `c`'s table id and column id despite being munged up by all of these filter and join operators.

**Review by jordanlewis — APPROVED** [review]
LGTM overall. I just am not sure if this is "complete" yet, and adding some more tests would help find out. Nonetheless even if we didn't make it "complete" in this way it would still be worth merging.

File: pkg/sql/pgwire/testdata/pgtest/int_size (hunk 1 above)
Comment (rafiss) [reply pkg/sql/pgwire/testdata/pgtest/int_size:95]: sure thing!

**Comment by rafiss** [comment]
This is ready for another review. The additional tests were helpful in finding another place to pass through this metadata.

**Review by jordanlewis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48417)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[pkg/sql/copy.go, line 137 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48417#-M6gyBpw7RhjNeI1aHmb:-M6gyBpw7RhjNeI1aHmc:b-22frgt) ([raw file](https://github.com/cockroachdb/cockroach/blob/459ec99950ce6aa57e18db4586340ee7bdca597d/pkg/sql/copy.go#L137)):*
> ```Go
> 			Typ:      cols[i].Type,
> 			TableID:  tableDesc.GetID(),
> 			ColumnID: cols[i].ID}
> ```

Ooh, just realizing that these all need to be `.GetLogicalColumnID()`, since that's how `attnum` is defined now. 
___
*[pkg/sql/opt_exec_factory.go, line 1977 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48417#-M6h2wkO6KbxyystNeuZ:-M6h2wkO6KbxyystNeu_:boiuhyj) ([raw file](https://github.com/cockroachdb/cockroach/blob/459ec99950ce6aa57e18db4586340ee7bdca597d/pkg/sql/opt_exec_factory.go#L1977)):*
> ```Go
> // addExpr adds a new render expression with the given name.
> func (rb *renderBuilder) addExpr(
> 	expr tree.TypedExpr, colName string, tableID sqlbase.ID, columnID sqlbase.ColumnID,
> ```

If you change the other one, columnID should also change to be called pgAttributeNum here.
___
*[pkg/sql/pgwire/conn.go, line 1287 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48417#-M6gz1rN3nYsQousxEEu:-M6gxdZU3T9vWFmsmzKf:b9cf88q) ([raw file](https://github.com/cockroachdb/cockroach/blob/459ec99950ce6aa57e18db4586340ee7bdca597d/pkg/sql/pgwire/conn.go#L1287)):*
> ```Go
> 		typ := pgTypeForParserType(column.Typ)
> 		c.msgBuilder.putInt32(int32(column.TableID))  // Table OID (optional).
> 		c.msgBuilder.putInt16(int16(column.ColumnID)) // Column attribute ID (optional).
> ```

This downcast is scary. I think it means that if you churn through 65k add/drops it'll break things silently...

@richardjcai @rohany do we have limits on this anywhere that you know of? Since you've been looking at this stuff.
___
*[pkg/sql/pgwire/testdata/pgtest/row_description, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48417#-M6h2Zzz-vWZJP5XCH1b:-M6h2Zzz-vWZJP5XCH1c:b-2lmmy7) ([raw file](https://github.com/cockroachdb/cockroach/blob/459ec99950ce6aa57e18db4586340ee7bdca597d/pkg/sql/pgwire/testdata/pgtest/row_description#L1)):*
> ```
> # This test verifies that we're populating the TableID and ColumnID in the
> ```

s/ColumnID/PGAttributeNum/
___
*[pkg/sql/pgwire/testdata/pgtest/row_description, line 80 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48417#-M6h3Tb9AbM5GayyV4UM:-M6h3Tb9AbM5GayyV4UN:b10ca2m) ([raw file](https://github.com/cockroachdb/cockroach/blob/459ec99950ce6aa57e18db4586340ee7bdca597d/pkg/sql/pgwire/testdata/pgtest/row_description#L80)):*
> ```
> 
> send
> Query {"String": "SELECT * FROM v WHERE v1 = 1"}
> ```

Nice tests!
___
*[pkg/sql/sqlbase/result_columns.go, line 32 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48417#-M6gy_MpASYuKGKRsspZ:-M6gy_MpASYuKGKRssp_:b-lr8t3a) ([raw file](https://github.com/cockroachdb/cockroach/blob/459ec99950ce6aa57e18db4586340ee7bdca597d/pkg/sql/sqlbase/result_columns.go#L32)):*
> ```Go
> 	// reference, these fields are zeroes.
> 	TableID  ID       // OID of column's source table (pg_attribute.attrelid).
> 	ColumnID ColumnID // Column's number in source table (pg_attribute.attnum).
> ```

I think we should change this field to be called `PGAttributeNum`, since it's actually distinct from the idea of `ColumnID` as you note.
___
*[pkg/sql/sqlbase/result_columns.go, line 58 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48417#-M6gyn_I6opG3Hf4BqZj:-M6gyn_I6opG3Hf4BqZk:b7m5dwj) ([raw file](https://github.com/cockroachdb/cockroach/blob/459ec99950ce6aa57e18db4586340ee7bdca597d/pkg/sql/sqlbase/result_columns.go#L58)):*
> ```Go
> 				Hidden:   hidden,
> 				TableID:  tableID,
> 				ColumnID: colDesc.ID,
> ```

ditto: This should be colDesc.GetLogicalColumnID(), how `attnum` is defined.

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48417)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss, @RichardJCai, and @rohany)

___
*[pkg/sql/pgwire/conn.go, line 1287 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48417#-M6gz1rN3nYsQousxEEu:-M6h4UGlDqa42FmbimPy:b-fqgu51) ([raw file](https://github.com/cockroachdb/cockroach/blob/459ec99950ce6aa57e18db4586340ee7bdca597d/pkg/sql/pgwire/conn.go#L1287)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

This downcast is scary. I think it means that if you churn through 65k add/drops it'll break things silently...

@richardjcai @rohany do we have limits on this anywhere that you know of? Since you've been looking at this stuff.
</blockquote></details>

it shouldn't actually be dangerous, because `pg_attribute.attnum` is an `INT2` type

but i agree, the types make it scary. `ColumnID` should be an alias for int16

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48417)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @rafiss, @RichardJCai, and @rohany)

___
*[pkg/sql/copy.go, line 137 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48417#-M6gyBpw7RhjNeI1aHmb:-M6h5FH5DWrePIBDu1Q3:bzvuo2) ([raw file](https://github.com/cockroachdb/cockroach/blob/459ec99950ce6aa57e18db4586340ee7bdca597d/pkg/sql/copy.go#L137)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Ooh, just realizing that these all need to be `.GetLogicalColumnID()`, since that's how `attnum` is defined now. 
</blockquote></details>

nice catch, i'll update them

**Review by rohany — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48417)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @rafiss, and @RichardJCai)

___
*[pkg/sql/pgwire/conn.go, line 1287 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48417#-M6gz1rN3nYsQousxEEu:-M6h9qI6CY2mf6lr1rLI:b-bu9fcz) ([raw file](https://github.com/cockroachdb/cockroach/blob/459ec99950ce6aa57e18db4586340ee7bdca597d/pkg/sql/pgwire/conn.go#L1287)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

it shouldn't actually be dangerous, because `pg_attribute.attnum` is an `INT2` type

but i agree, the types make it scary. `ColumnID` should be an alias for int16
</blockquote></details>

I don't believe that we have any tracked limits on this -- maybe we need to start keeping track

**Review by jordanlewis — APPROVED** [review]
:lgtm: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48417#-:-M6kTxW9BwqYORDx_OgZ:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rafiss)

___
*[pkg/sql/opt_exec_factory.go, line 1977 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48417#-M6h2wkO6KbxyystNeuZ:-M6kTjDR-FTxQgmcXW-W:b-stkect) ([raw file](https://github.com/cockroachdb/cockroach/blob/459ec99950ce6aa57e18db4586340ee7bdca597d/pkg/sql/opt_exec_factory.go#L1977)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

If you change the other one, columnID should also change to be called pgAttributeNum here.
</blockquote></details>

Heh, I see that GetLogicalColumnID returns sqlbase.ColumnID which feels wrong. I don't think you have to change in this PR but it's a little weird.

**Comment by rafiss** [comment]
thanks for the reviews!

bors r+

**Pull request merged**